### ISSUE TITLE
Feature/Functionality Change Request: ObjectCreationHandling.Replace as Default Behavior

### ISSUE THREAD

**Issue by zdougherty18** [JamesNK/Newtonsoft.Json#1535]
Solution/Feature Requested: ObjectCreationHandling.Replace should be the default behavior when deserializing in place of ObjectCreationHandling.Reuse.

Issue #486 and #703 appear to be dealing with the same problem concept as I am requesting a solution for. I bring them up as I believe they are integral to the common use case and potential fallout from the logical functionality when dealing with computed-properties. For the sake of clarity, computed-properties are readonly on the scope of the full object and therefor not assignable on that level, but each of the sub-properties are assignable based on their own definition. Also, deserialization is the process of taking supplied raw data and constructing a model based off of the closest matching definition.

Now, when we take data and attempt to map out an object; there is little to no knowledge of what may have happened while that data was "in the wild" or how the data may have been altered. So, one would assume you must verify the data-structure while consuming that data. Also, as we can see by the Json.NET library, you may want to alter that data-structure on import, which can be accomplished by attributes and other serialization settings.

With all of this in mind, we have to ask ourselves how we should handle a readonly property while deserializing and mapping the data. I believe in general, one should say that the most intuitive and probable use-case would be; readonly properties and their sub properties should be ignored. 

When we are mapping data we are generally trying to map a whole structure and would be prudent to treat it as such with a top-down approach. It seems contradictory to attempt to copy the object over, get denied, and then take each of its sub-properties and copy them over anyways. While this is technically allowed by the definition of populating object's class; I think it is logically inferior as we ignore the full known state of the data coming in while hydrating the object. With that said, I also recognize the use case of having a computed-property for deserialization and expecting those computed-property's properties to be set. I would still say this is an inferior use case.

Also, I think this would be the most likely conclusion one would come to while consuming a deserializer. With that in mind, if it behaves differently than the most common logical conclusion I would say we haven't used the best case scenario and have allowed for preventable data loss. The only reason I bring this up is that is what happened to us, and I believe any public product should try to minimize failure by defaulting to the best case. 

In addition, not mapping readonly properties or their sub-properties would stay consistent with the default
_System.Web.Script.Serialization.JavaScriptSerializer_ functionality set by Microsoft.

For LINQPad(just change the serializer to Newtonsoft's to compare:

```
void Main()
{
	Compound theObject = new Compound() { MySimple = new Simple() { ValOne=1, ValTwo=2 }};
	theObject.Dump();
	
	System.Web.Script.Serialization.JavaScriptSerializer serializer = new System.Web.Script.Serialization.JavaScriptSerializer();
	
	string json = serializer.Serialize(theObject);
	json.Dump();	
	
	"\n----This is where we set the changed json values on the equivelant of the Computed Property-----".Dump();
	json = "{\"MySimple\":{\"ValOne\":1,\"ValTwo\":2},\"SimpleComputation\":{\"ValOne\":5,\"ValTwo\":5}}";
	json.Dump();
	"\n".Dump();
	
	theObject = null;
	theObject.Dump();
	
	theObject = serializer.Deserialize<Compound>(json);
	theObject.Dump();
}

public class Compound 
{
	public Simple MySimple {get;set;}
	public Simple SimpleComputation {get {return MySimple;}}
}

public class Simple
{
	public int ValOne {get;set;}
	public int ValTwo {get;set;}
}
```

**Comment by JamesNK** [comment]
I agree, I would change it but it will break a lot of applications that depend on the existing behavior. It is too big a breaking change to ever make.

**Comment by zdougherty18** [comment]
You are absolutely right, and and breaking functionality changes should not be implemented quick. It is definitely not something to be gone to lightly. I would like to offer a counter solution though. On the next major revision release, which would theoretically have other breaking changes, could it be updated at that time with release notes being pushed out. 

At that point it would be on the consumers of the API to make sure their code works as intended.

**Comment by TylerBrinkley** [comment]
That's not going to happen. Due to the popularity of this library as a dependency of other libraries @JamesNK tries to avoid breaking changes even on major version releases.

**Comment by zdougherty18** [comment]
You lock yourself into a bad functionality because you don't want to ever introduce breaking changes? That is both inflexible and not how software development is supposed to work. Also, it is exactly what major version releases are supposed to be.

Here's a bit on semantic versioning: [https://semver.org/](https://semver.org/)
